File: solutions/java/beer-song/1/src/main/java/BeerSong.java
import java.util.stream.IntStream;

import static java.util.stream.Collectors.joining;

public class BeerSong {

    public static String verse(int n) {
        String bottleCount = bottles(n);
        return uppercaseFirstLetter(bottleCount) + " of beer on the wall, "
                + bottleCount + " of beer.\n"
                + action(n)
                + ", " + bottles(n - 1) + " of beer on the wall.\n\n";
    }

    public static String sing(int from, int to) {
        return IntStream.rangeClosed(to, from)
                        .map(i -> to - i + from)
                        .mapToObj(BeerSong::verse)
                        .collect(joining(""));
    }

    public static String singSong() {
        return sing(99, 0);
    }

    private static String bottles(int n) {
        switch (n) {
            case 0: return "no more bottles";
            case 1: return "1 bottle";
            case -1: return "99 bottles";
            default: return n + " bottles";
        }
    }

    private static String uppercaseFirstLetter(String str) {
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }

    private static String action(int n) {
        switch (n) {
            case 0: return "Go to the store and buy some more";
            case 1: return "Take it down and pass it around";
            default: return "Take one down and pass it around";
        }
    }
}

File: solutions/erlang/beer-song/1/beer_song.erl
-module(beer_song).
-export([verse/1, sing/1, sing/2]).

verse(Bottles) ->
  current_bottles(Bottles) ++ next_action(Bottles).

sing(Start) ->
  sing(Start, 0).

sing(Start, End) ->
  sing(Start, End, []).

sing(Start, End, Acc) when Start < End ->
  Acc;
sing(Start, End, Acc) ->
  sing(Start - 1, End, Acc ++ verse(Start) ++ "\n").

current_bottles(0) ->
  "No more bottles of beer on the wall, no more bottles of beer.\n";
current_bottles(1) ->
  "1 bottle of beer on the wall, 1 bottle of beer.\n";
current_bottles(Bottles) ->
  io_lib:fwrite("~p bottles of beer on the wall, ~p bottles of beer.~n", [Bottles, Bottles]).

next_action(0) ->
  "Go to the store and buy some more, 99 bottles of beer on the wall.\n";
next_action(1) ->
  "Take it down and pass it around, no more bottles of beer on the wall.\n";
next_action(2) ->
  "Take it down and pass it around, 1 bottle of beer on the wall.\n";
next_action(Bottles) ->
  io_lib:fwrite("Take it down and pass it around, ~p bottles of beer on the wall.~n", [Bottles - 1]).